uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -116,476 +76,627 @@
+    // The UniFfiTag ZST. The Send + 'static bound is to keep rustc happy.
+    UT: Send + 'static,
+{
+    // Create a RustFuture and coerce to `Arc<dyn RustFutureFFI>`, which is what we use to
+    // implement the FFI
+    let future_ffi = RustFuture::new(future, tag) as Arc<dyn RustFutureFFI>;
+    // Box the Arc, to convert the wide pointer into a normal sized pointer so that we can pass it
+    // to the foreign code.
+    let boxed_ffi = Box::new(future_ffi);
+    // We can now create a RustFutureHandle
+    RustFutureHandle(Box::into_raw(boxed_ffi) as *mut ())
+}
+
+/// Poll a Rust future
+///
+/// When the future is ready to progress the continuation will be called with the `data` value and
+/// a [RustFuturePoll] value. For each [rust_future_poll] call the continuation will be called
+/// exactly once.
 ///
-/// The foreign code passes a pointer to one of these callbacks along with an opaque data pointer.
-/// When the future is ready, we invoke the callback.
-pub type FutureCallback<T> =
-    extern "C" fn(callback_data: *const (), result: T, status: RustCallStatus);
+/// # Safety
+///
+/// The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+pub unsafe fn rust_future_poll(handle: RustFutureHandle, data: *const ()) {
+    let future = &*(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    future.clone().ffi_poll(data)
+}
 
-/// Future that the foreign code is awaiting
+/// Cancel a Rust future
+///
+/// Any current and future continuations will be immediately called with RustFuturePoll::Ready.
+///
+/// This is needed for languages like Swift, which continuation to wait for the continuation to be
+/// called when tasks are cancelled.
 ///
-/// RustFuture is always stored inside a Pin<Arc<>>.  The `Arc<>` allows it to be shared between
-/// wakers and Pin<> signals that it must not move, since this would break any self-references in
-/// the future.
-pub struct RustFuture<F, T, UT>
+/// # Safety
+///
+/// The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+pub unsafe fn rust_future_cancel(handle: RustFutureHandle) {
+    let future = &*(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    future.clone().ffi_cancel()
+}
+
+/// Complete a Rust future
+///
+/// Note: the actually extern "C" scaffolding functions can't be generic, so we generate one for
+/// each supported FFI type.
+///
+/// # Safety
+///
+/// - The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+/// - The `T` param must correctly correspond to the [rust_future_new] call.  It must
+///   be `<Output as FfiConverter<UT>>::ReturnType`
+pub unsafe fn rust_future_complete<T: FfiDefault>(
+    handle: RustFutureHandle,
+    out_status: &mut RustCallStatus,
+) -> T {
+    let future = &*(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    let mut return_value = T::ffi_default();
+    let out_return = mem::transmute::<&mut T, &mut ()>(&mut return_value);
+    future.ffi_complete(out_return, out_status);
```

Comment:
This looks very wrong. Why are safe reference types being used here?

Reply:
I don't understand this one.  Both `rust_future_complete` and `ffi_complete` are unsafe functions precisely because they transmute the references.  The safety requirements are listed above.  If those are followed, I think the code should work because you'll end up with something like:
```
let mut return_value: T::ReturnType = T::ffi_default();
let out_return = mem::transmute::<&mut T, &mut ()>(&mut return_value);
let out_return = std::mem::transmute::<&mut (), &mut T::ReturnType>(out_return);
```

That should be okay right?

Reply:
That's very interesting, although I still don't follow it all.  Wouldn't aliasing need to involve 2 pointers/references?  Is the issue that some of these transmutes are copying the `&ref` so you end up with 2 of them?

RE provenance: does it work differently between pointers and references?  The RawWaker API casts `*const T` to `*const ()`, so I figured I could do the same with references.  My rough mental model is that provenance is set when the pointer is allocated and sticks around through a cast/transmute.

In any case, all of this is making me want to rework things we keep the generic `T` around and avoid all of these issues.  I think that means we need to define a poll/cancel/complete/free scaffolding function for each exported async function.  That sounds a bit sad, but worth it keep things simple.

Reply:
> That's very interesting, although I still don't follow it all. Wouldn't aliasing need to involve 2 pointers/references? Is the issue that some of these transmutes are copying the `&ref` so you end up with 2 of them?

Oh, right transmute moves the `&mut` so maybe it is fine in that regard.

> RE provenance: does it work differently between pointers and references? The RawWaker API casts `*const T` to `*const ()`, so I figured I could do the same with references. My rough mental model is that provenance is set when the pointer is allocated and sticks around through a cast/transmute.

Yes, provenance is different between pointers and references, at least under stacked borrows. I believe it's different with the tree borrows model. Under stacked borrows, AFAIK a (non-DST) reference only ever has provenance over as many bytes as there are in the type it references, so `&mut ()` has provenance over nothing. Pointers retain provenance when cast.

> In any case, all of this is making me want to rework things we keep the generic `T` around and avoid all of these issues. I think that means we need to define a poll/cancel/complete/free scaffolding function for each exported async function. That sounds a bit sad, but worth it keep things simple.

That's what Ivan's initial implementation did, right?

Reply:
Indeed, that was part of my initial implem', and also part of my last PR.